File: de.unistuttgart.iste.sqa.mpw.hamstersimulator.cpp/core/src/impl/AddEntityCommandImpl.cpp
#include "AddEntityCommandImpl.h"

#include "Entity.h"

namespace commands {

AddEntityCommandImpl::AddEntityCommandImpl() {
}

void AddEntityCommandImpl::execute() {
    auto entity = getEntity();
    EntityReference reference = getEntityToAdd().get();
    entity->addToCollection(getFeatureKey(), reference);
}

void AddEntityCommandImpl::undo() {
    auto entity = getEntity();
    EntityReference reference = getEntityToAdd().get();
    entity->removeFromCollection(getFeatureKey(), reference);
}

void AddEntityCommandImpl::redo() {
	execute();
}

}

File: de.unistuttgart.iste.sqa.mpw.karasimulator.java/core/src/main/java/de/unistuttgart/kara/commands/impl/AddEntityCommandImpl.java
package de.unistuttgart.kara.commands.impl;

import de.unistuttgart.kara.commands.AddEntityCommand;
import de.unistuttgart.kara.commands.impl.utils.CommandUtils;

public class AddEntityCommandImpl extends AddEntityCommand {

	public AddEntityCommandImpl() {

	}

	@Override
	public void execute() {
		try {
			var entity = getEntity();
			var method = CommandUtils.findAddMethod(entity, getPropertyName());
			method.invoke(entity, getEntityToAdd());
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	@Override
	public void undo() {
		try {
			var entity = getEntity();
			var method = CommandUtils.findRemoveMethod(entity, getPropertyName());
			method.invoke(entity, getEntityToAdd());
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	@Override
	public void redo() {
		execute();
	}

}
